### DISCUSSION TITLE
Type-safe "RPC" Client

### DISCUSSION THREAD

**Discussion by joe-hillyard** [TypedProject/ts-express-decorators#2289] [Q&A]
I recently saw an enhancement issue for tRPC integration and while it is an interesting concept I'm not sure i'm willing to migrate my projects to tRPC in any capacity, but it did cause me to think about type-safe RPC and how it could realistically be achieved within Ts.ED.

For context, I am am using [moon repo](https://moonrepo.dev/) as my monorepo with several backend REST services, api gateway and multiple user interfaces as one whole workspace.  However one of my biggest issues I haven't been able to solve yet is isomorphic/homogenous clients for a streamlined developer experience/workspace integration.

I stumbled across [this article](https://trpc.io/blog/tinyrpc-client) which highlights the typescript magic and JS proxy internals required to make it achievable, therefore my question is - do you have any suggestions for a good place for me to start investigating? I don't see much documented on PlatformRouter but I imagine that would be the best point of integration.

Thanks in advance!

**Comment by joe-hillyard** [comment]
Perhaps I did not do enough research before posting - I have just found the ["generate-http-client" CLI command](https://github.com/tsedio/tsed-cli/blob/master/packages/cli-generate-http-client/src/commands/GenerateHttpClientCmd.ts). This will help narrow down the problem (I have been using [orval](https://orval.dev/) up until now, but would prefer to bring internally).

I do notice that this command is resolving the framework platform (express or koa) then bootstrapping etc. My understanding was the Ts.ED Platform Router is an abstraction that can be consumed the same way, or have I misinterpreted?

**Reply by Romakita** [reply]
Hello @joe-hillyard 

I'm really sorry for my late answer. You've totally right, the documentation is missing around the platform router and I try to develop this part by adding a new documentation page that explain how any developer can implement his own Platform Adapter.

The Platform Router is just an abstraction and doesn't intend to implement an entire Router (the routing resolution isn't implemented by Ts.ED). His concern, is to collect the routes configured on each controllers (you can find a small readme that explain the principle: https://www.npmjs.com/package/@tsed/platform-router).

The generate-http-client, get the Server class and bootstrap the server (without listening port) to get the swagger spec. The swagger spec reflect all routes registered by controllers and the mount options. Then, the CLI use a third party module to generate the code.

The OpenSpec aka Swagger / JsonSchema is the core of the Ts.ED framework and are implemented in the `@tsed/schema` module. `@tsed/platform-routers` consume the metadata stored by `@tsed/schema` to produce Layers (layers is a collection of HTTP Verb + Methods + Handler). Layers can be mapped to any server framework like Express or Koa.

I hope my explanation is clear ;) 

I'll try to give you more detail on the platform router soon (the doc is under worker in progress). Until that, we can continue to discuss about that here.

See you
Romain

**Reply by Romakita** [reply]
Here a document (wip) about adapters and platform-routers:
https://github.com/tsedio/tsed/blob/583e20c76ee83f5a433ba0dffba403cf3c3ca631/docs/docs/create-platform-adapter.md

**Reply by joe-hillyard** [reply]
@Romakita No need for apologies, you're a busy guy and I appreciate you will respond when you have time :) This information is very useful - i'm going to read further and try to build a proof-of-concept. Will keep this thread updated!

**Reply by Romakita** [reply]
@joe-hillyard have you get a time to work on it recently ?
See you ;)
